WT-15614 Performance regression due to missing optimisation flags for Release builds

ablekhman opened the pull request:
- Fix: missing optimisation flags for `Release`|`RelWithDebInfo` build types
- Optimisation flags were defined after they have been used.

ablekhman committed:
WT-15614 Performance regression
- Fix: missing optimisation flags for `Release`|`RelWithDebInfo` build types
- Optimisation flags were defined after they have been used.

ablekhman commented on cmake/configs/base.cmake:
```diff
@@ -459,6 +462,39 @@ if(ENABLE_DEBUG_INFO AND NOT WT_DEBUG_FLAGS_INITIALIZED)
 endif()
 
+# We want to use the optimization level from CC_OPTIMIZE_LEVEL.
+if(NOT ("${WT_OPTIMIZE_FLAGS_SAVED}" STREQUAL "${CC_OPTIMIZE_LEVEL}"))
```
We need to save previous value of `CC_OPTIMIZE_LEVEL` in separate variable `WT_OPTIMIZE_FLAGS_SAVED`.
If user performs configure, then edits `CC_OPTIMIZE_LEVEL` and configures again, then we need to update relevant variables with compiler flags. But only if the new value of the `CC_OPTIMIZE_LEVEL` is different from the previous one.

ablekhman commented on cmake/configs/base.cmake:
```diff
@@ -459,6 +462,39 @@ if(ENABLE_DEBUG_INFO AND NOT WT_DEBUG_FLAGS_INITIALIZED)
         "WiredTiger debug flags have been initialized")
 endif()
 
+# We want to use the optimization level from CC_OPTIMIZE_LEVEL.
```
Both debug and optimisation flags are now set in base.cmake. Previously, some of them were set in base.cmake and others in modes.cmake. This caused the bug when `CC_OPTIMIZE_LEVEL` was used in modes.cmake before it was properly set in base.cmake

ablekhman commented on cmake/helpers.cmake:
```diff
@@ -709,15 +709,19 @@ function(replace_compile_options flag_var)
     # Remove existing flags
     foreach(flag ${REPLACE_REMOVE})
-        string(REPLACE "${flag}" "" ${flag_var} ${${flag_var}})
+        string(REPLACE " ${flag} " "" ${flag_var} " ${${flag_var}} ")
```
Fixing a bug when removing a flag breaks unrelated flags.
Example: Removing "-g" from the string "-g -g3 -ggdb" would produce result: "3 gdb".

ivankochin replied:
Could you please add a comment explaining why do we need these spaces around flags

Siddhartha8899 commented:
Sorry don't have capacity to review this week. 
But 2 nits - 
- Changing the title of the PR to something descriptive
- I have run performance patches and shared with you @ablekhman

ivankochin commented on cmake/configs/base.cmake:
```diff
@@ -46,7 +48,8 @@ if(NOT HAVE_BUILTIN_EXTENSION_IAA)
 endif()
 
-if("${CMAKE_BUILD_TYPE}" STREQUAL "Release")
+
```
Extra blank line

ivankochin commented on cmake/helpers.cmake:
```diff
@@ -709,15 +709,19 @@ function(replace_compile_options flag_var)
+        string(REPLACE " ${flag} " "" ${flag_var} " ${${flag_var}} ")
     endforeach()
 
+    # Clean up extra spaces
+    string(STRIP "${${flag_var}}" ${flag_var})
```
Can we cleanup extra spaces only once in the end of the function?

ablekhman replied:
We have to trim the string twice because after flags are removed, there might be extra spaces at the end. If we append a new flag without removing tail spaces, then extra spaces will be trapped in the middle of the string.

ablekhman committed:
WT-15614 Performance regression due to missing optimisation flags
- Fix: missing optimisation flags for `Release`|`RelWithDebInfo` build types
- Optimisation flags were defined after they have been used.

ivankochin reviewed (approved):
LGTM!
